pyqtorch/primitive.py

```diff
@@ -1,128 +1,48 @@
 from __future__ import annotations
 
-import logging
 from functools import cached_property
-from logging import getLogger
 from typing import Any
 
-import numpy as np
 import torch
 from torch import Tensor
 
-from pyqtorch.apply import apply_operator, operator_product
 from pyqtorch.embed import Embedding
-from pyqtorch.matrices import OPERATIONS_DICT, _controlled, _dagger
-from pyqtorch.utils import DensityMatrix, expand_operator, permute_basis, product_state
+from pyqtorch.matrices import OPERATIONS_DICT, _controlled
+from pyqtorch.quantum_ops import QuantumOperation, Support
+from pyqtorch.utils import (
+    product_state,
+    qubit_support_as_tuple,
+)
 
-logger = getLogger(__name__)
 
+class Primitive(QuantumOperation):
+    """Primitive are fixed quantum operations with a defined unitary operator U.
 
-def forward_hook(*args, **kwargs) -> None:  # type: ignore[no-untyped-def]
-    torch.cuda.nvtx.range_pop()
 
+    Attributes:
+        operation (Tensor): Unitary tensor U.
+        qubit_support: List of qubits the QuantumOperation acts on.
+        generator (Tensor): A tensor G s.t. U = exp(-iG).
+    """
```

Comment:
I would say Primitives are operators based on some fixed matrix. It does not need to be unitary.